src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs

```diff
@@ -253,6 +253,68 @@ public async Task CanHandleRequestsWithUrlInAbsoluteForm(string requestUrl, stri
             }
         }
 
+        [Fact]
+        public async Task ExecutionContextDoesNotLeakAcrossRequestsOnSameConnection()
+        {
+            var local = new AsyncLocal<int>();
+
+            // It's important this method isn't async as that will revert the ExecutionContext
+            Task ExecuteApplication(HttpContext context)
+            {
+                var value = local.Value;
+                Assert.Equal(0, value);
+                local.Value++;
+
+                context.Response.OnStarting(() =>
+                {
+                    local.Value++;
+                    return Task.CompletedTask;
+                });
+
+                context.Response.OnCompleted(() =>
+                {
+                    local.Value++;
+                    return Task.CompletedTask;
+                });
+
+                context.Response.ContentLength = 1;
+                return context.Response.WriteAsync(value.ToString());
+            }
```

Comment:
I've been playing with this test in this branch to see how these changes affect the ExecutionContext, but even after these changes the behavior is kinda weird IMHO. I was taking a look at the actual value of `local.Value` in the OnStarting and OnCompleted callbacks, and it doesn't make much sense to me.

The following version of ExecuteApplication still passes the test with these changes:

```C#
Task ExecuteApplication(HttpContext context)
{
    var value = local.Value;
    Assert.Equal(0, value);
    local.Value++;

    context.Response.OnStarting(() =>
    {
        Assert.Equal(1, local.Value);
        local.Value++;
        return Task.CompletedTask;
    });

    context.Response.OnCompleted(() =>
    {
        Assert.Equal(0, local.Value);
        local.Value++;
        return Task.CompletedTask;
    });

    context.Response.ContentLength = 1;
    return context.Response.WriteAsync(value.ToString());
}
```

Notice that `local.Value` is `1` at the start of OnStarting callback, but reverts back to `0` by the time the OnCompleted callback is called.

I think that if `local.Value` was reverted to `0` in the OnStarting callback *or* if `local.Value` was not reverted in either callback and remained `2` at the start of the OnCompleted callback, it would be more consistent/logical.

Reply:
One other option that might be less weird would be to reset after the entire request. I looked at that but it felt overly complicated

Reply:
@anurse and I talked about this. We think this behavior is the result of `OnStarting` being invoked inside of `application.ProcessRequestAsync` via `Response.WriteAsync`, whereas `OnCompleted` gets called after `application.ProcessRequestAsync` exits ignoring a possible call to `Response.PipeWriter.CompleteAsync()` or something like that.

I don't think there's much we can do to fix that.

Reply:
It might still be worth adding those extra asserts in the OnStarting/OnCompleted callbacks so we don't unintentionally change/regress behavior around this though.

Reply:
Hmmm, this might actually be problematic if you expect to access async locals in OnStarting without having written anything (though that might be rare). I need to run some tests with the IHttpContextAccessor to see how it affects those callbacks. This is basically a breaking change

Reply:
> One other option that might be less weird would be to reset after the entire request. I looked at that but it felt overly complicated

Wasn't sure if that would be correct behaviour; but it should be simplier. Give me a few mins will make a PR so you can see what it looks like.

Reply:
https://github.com/aspnet/AspNetCore/pull/14146/files?w=1

Reply:
That adds more per request allocation in other cases. I was trying to avoid that.